dali/kernels/algebra/linear_transformation_gpu.h

```diff
@@ -0,0 +1,198 @@
+  OutputType *out;
+  ivec<spatial_ndims> in_size, in_strides, out_size, out_strides;
+  int in_channels, out_channels;
+  mat<N, M, TransformationMatrixElementType> transformation_matrix;
+};
+
+
+template <class OutputType, class InputType, size_t N, size_t M, size_t spatial_ndims>
+std::vector<SampleDescriptor<OutputType, InputType, N, M, spatial_ndims>>
+CreateSampleDescriptors
+        (const OutListGPU<OutputType, spatial_ndims + 1> &out,
+         const InListGPU<InputType, spatial_ndims + 1> &in,
+         const std::vector<mat<N, M, TransformationMatrixElementType>> &transformation_matrices) {
+  std::vector<SampleDescriptor<OutputType, InputType, N, M, spatial_ndims>> ret(
+          in.num_samples());
+
+  for (int i = 0; i < in.num_samples(); i++) {
+    auto &sample = ret[i];
+    sample.in = in[i].data;
+    sample.out = out[i].data;
+
+    auto get_size = [](const TensorShape<spatial_ndims + 1> &ts) -> auto {
+        ivec<spatial_ndims> ret;
+        for (int i = ret.size() - 1, j = 0; i >= 0; i--, j++) {
+          ret[j] = ts[i];
+        }
+        return ret;
+    };
+
+    auto get_channels = [](const TensorShape<spatial_ndims + 1> &ts) -> auto {
+        return ts.shape.back();
+    };
+
+    sample.in_size = get_size(in.tensor_shape(i));
+    sample.out_size = get_size(out.tensor_shape(i));
+    sample.in_channels = get_channels(in.tensor_shape(i));
+    sample.out_channels = get_channels(out.tensor_shape(i));
+    sample.in_strides = {sample.in_channels, sample.in_size.x * sample.in_channels};
+    sample.out_strides = {sample.out_channels, sample.out_size.x * sample.out_channels};
+    sample.transformation_matrix = transformation_matrices[i];
+  }
+
+  return ret;
+}
+
+
+}  // namespace detail
+
+
+template <class OutputType, class InputType, size_t N, size_t M, size_t spatial_ndims>
+void __global__ LinearTransformationKernel(
+        const detail::SampleDescriptor<OutputType, InputType, N, M, spatial_ndims> *samples,
+        const BlockDesc<spatial_ndims> *blocks) {
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  const Surface2D<const InputType> in = {
+          sample.in, sample.in_size.x, sample.in_size.y, sample.in_channels,
+          sample.in_strides.x, sample.in_strides.y, 1
+  };
+
+  const Surface2D<OutputType> out = {
+          sample.out, sample.out_size.x, sample.out_size.y, sample.out_channels,
+          sample.out_strides.x, sample.out_strides.y, 1
+  };
+
+  for (int y = threadIdx.y + block.start.y; y < block.end.y; y += blockDim.y) {
+    for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+      // Iterate over channels: `o` over output, `i` over input
+      // It's a plain matrix multiplication
+      for (int o = 0; o < N; o++) {
+        OutputType val = 0;
+        for (int i = 0; i < M; i++) {
+          val += sample.transformation_matrix.at(o, i) * in(x, y, i);
+        }
+        out(x, y, o) = val;
```

Comment:
I don't know if you need to reimplement the matrix - vector multiplication.
Maybe something like that:
```c++
auto *in_vec = reinterpret_cast<vec<InputType, M>*>(&in(x, y, 0));
auto *out_vec =  reinterpret_cast<vec<OutputType, N>*>(&out(x, y, 0));
*out_vec = sample.transformation_matrix * (*in_vec);
```
Or instead of these reinterpret_casts you could just copy memory to temporary vecs.

Reply:
@banasraf 
1. This would only work with dense HWC layout.
2. We can reinterpret input, but for the output we need to apply proper conversion anyway.